src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/NoMetadataBlockingPolicy.cs

```diff
@@ -34,13 +44,22 @@ public override bool IsBlocked(MethodDesc method)
         {
             if (method is EcmaMethod ecmaMethod)
             {
+                TypeDesc owningType = ecmaMethod.OwningType;
+
                 // Methods on Array`1<T> are implementation details that implement the generic interfaces on
                 // arrays. They should not generate metadata or be reflection invokable.
                 // We can get rid of this special casing if we make these methods stop being regular EcmaMethods
                 // with Array<T> as their owning type
-                if (ecmaMethod.OwningType == GetArrayOfTType(ecmaMethod))
+                if (owningType == GetArrayOfTType(ecmaMethod))
                     return true;
 
+                // Do not expose any of our internal details
+                if (owningType.IsObject)
+                {
+                    return method.Name is not ".ctor" and not "Equals" and not "Finalize" and not "GetHashCode"
+                        and not "GetType" and not "MemberwiseClone" and not "ReferenceEquals" and not "ToString";
```

Comment:
Would it be better to check whether the method is public instead?

Reply:
`Finalize` and `MemberwiseClone` are not public and we change the non-public surface are so often that it feels like the public surface area is more stable.

Reply:
Ok, I should have said visible outside the assembly (ie public, family, FamOrAssembly).

>  we change the non-public surface are so often 

I agree that we are iterating on the non-public (ie not visible outside the assembly) surface frequently. However, we try to avoid adding non-standard surface visible outside the assembly to otherwise public types. It requires adding exceptions to the libraries partitions. 

Hardcoding a list like this into the compiler feels dirty.

Reply:
I'll look into what it takes to simply get rid of those implementation details. It goes towards the goal of unifying the CoreLibs across the JIT and no-JIT flavors.